### PR TITLE
Nerf ninja research stealing

### PR DESCRIPTION
**Pull request by EmoGarbage404**
## About the PR

Changes ninja gloves to steal a flat rate of techs instead of wiping all of them.
Additionally adjusts the objective to require 1 to 2 steals to reach the target amount.

## Why / Balance

Ninja's stealing research was previously an extremely trivial and annoying objective. For them, you can pretty easily break into the server room and wait out the 20 second doafter. It's not super hard and you're not exactly shooting off warning signs. At this very low cost, science loses literally an hour+ of progress.

This nerf changes it to only remove 4-8 random technologies, or only about 20k to 40k points worth. This is still pretty crippling but not awful. On top of that, they need to do it twice to hit the target amounts of 9 to 12 techs stolen. This provides counterplay by actually giving science a chance to see that all their research has vanished before the ninja is already gone. It also makes it significantly harder for ninja to obliterate everything science has done.

## Media

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- tweak: Ninjas no longer wipe all technologies when using their gloves on a research server.

### CODE REVIEW

**Comment by UbaserB** [comment]
Its so easy to fail it because science doesn’t research enough tech. Loving this change.

**Comment by deltanedas** [comment]
yeah just make it steal less and require less so you dont have to do scis job for green

**Comment by EmoGarbage404** [comment]
> Its so easy to fail it because science doesn’t research enough tech. Loving this change.

how early is ninja showing up that sci doesn't have 8 techs

**Comment by UbaserB** [comment]
> > Its so easy to fail it because science doesn’t research enough tech. Loving this change.
> 
> how early is ninja showing up that sci doesn't have 8 techs

they only had 3 techs, and everyone was dead on a warops round.

File: Content.Server/Research/Systems/ResearchStealerSystem.cs [hunk 1]
```diff
@@ -24,10 +26,20 @@ private void OnDoAfter(EntityUid uid, ResearchStealerComponent comp, ResearchSte
         if (!TryComp<TechnologyDatabaseComponent>(target, out var database))
             return;
 
-        var ev = new ResearchStolenEvent(uid, target, database.UnlockedTechnologies);
+        var ev = new ResearchStolenEvent(uid, target, new());
+        var count = _random.Next(comp.MinToSteal, comp.MaxToSteal + 1);
```
Comment (Tayrtahn) [inline Content.Server/Research/Systems/ResearchStealerSystem.cs:30]: But why +1?

File: Content.Server/Research/Systems/ResearchStealerSystem.cs [hunk 2]
```diff
@@ -24,10 +26,20 @@ private void OnDoAfter(EntityUid uid, ResearchStealerComponent comp, ResearchSte
...
+        for (var i = 0; i < count; i++)
+        {
+            if (database.UnlockedTechnologies.Count == 0)
+                break;
+
+            var toRemove = _random.Pick(database.UnlockedTechnologies);
+            _research.TryRemoveTechnology(toRemove, (target, database));
+            ev.Techs.Add(toRemove);
```
Comment (Tayrtahn) [inline Content.Server/Research/Systems/ResearchStealerSystem.cs:38]:
```suggestion
            if (_research.TryRemoveTechnology(toRemove, (target, database)))
               ev.Techs.Add(toRemove);
```
Should the ninja get credit for stealing even if the removal (somehow) fails?

File: Content.Shared/Research/Systems/SharedResearchSystem.cs [hunk 3]
```diff
@@ -219,16 +220,52 @@ public void TrySetMainDiscipline(TechnologyPrototype prototype, EntityUid uid, T
...
         Dirty(uid, component);
     }
 
+    public bool TryRemoveTechnology(ProtoId<TechnologyPrototype> tech, Entity<TechnologyDatabaseComponent> entity)
+    {
+        return TryRemoveTechnology(PrototypeManager.Index(tech), entity);
+    }
+
+    [PublicAPI]
+    public bool TryRemoveTechnology(TechnologyPrototype tech, Entity<TechnologyDatabaseComponent> entity)
```
Comment (Tayrtahn) [inline Content.Shared/Research/Systems/SharedResearchSystem.cs:236]: Maybe just personal preference, but wouldn't it be more consistent to have the Entity<> parameter first on these?

File: Content.Server/Research/Systems/ResearchStealerSystem.cs (hunk 2 above)
Comment (EmoGarbage404) [reply Content.Server/Research/Systems/ResearchStealerSystem.cs:38]: it shouldn't ever fail but sure

**Review by Tayrtahn — APPROVED** [review]

**Pull request merged**